### PR TITLE
drivers: dts: Add SPI capability to bmi270 driver

### PR DESCRIPTION
**Pull request by dchat-nordic**
Added SPI bus support for BMI270 gyroscope driver.

Signed-off-by: Dominik Chat <dominik.chat@nordicsemi.no>

### CODE REVIEW

**Review by galak — DISMISSED** [review]
Need to update drivers/sensor/bmi270/Kconfig:

```
        select I2C if $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BMI270),i2c)
        select SPI if $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BMI270),spi)
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please split the renaming/refactoring changes and introduction of SPI capability into two separate commits.

File: drivers/sensor/bmi270/bmi270.c [hunk 1]
```diff
@@ -584,17 +617,17 @@ static int bmi270_init(const struct device *dev)
...
 	uint8_t msg;
 	uint8_t tries;
 	uint8_t adv_pwr_save;
 
-	if (!device_is_ready(cfg->i2c.bus)) {
-		LOG_ERR("I2C bus device not ready");
-		return -ENODEV;
+	ret = bmi270_bus_check(dev);
+	if (ret < 0) {
+		LOG_ERR("Could not get pointer to %s device", dev->name);
```
Comment (mbolivar-nordic) [inline drivers/sensor/bmi270/bmi270.c:629]: Is this really the error you wish to log?

File: drivers/sensor/bmi270/bmi270.h [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
  * Copyright (c) 2021 Bosch Sensortec GmbH
+ * Copyright (c) 2022 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #ifndef ZEPHYR_DRIVERS_SENSOR_BMI270_BMI270_H_
 #define ZEPHYR_DRIVERS_SENSOR_BMI270_BMI270_H_
 
+
```
Comment (mbolivar-nordic) [inline drivers/sensor/bmi270/bmi270.h:11]: Remove unnecessary whitespace change

File: drivers/sensor/bmi270/bmi270.c [hunk 3]
```diff
@@ -545,13 +578,13 @@ static int bmi270_attr_set(const struct device *dev, enum sensor_channel chan,
 	    || (chan == SENSOR_CHAN_ACCEL_XYZ)) {
 		switch (attr) {
 		case SENSOR_ATTR_SAMPLING_FREQUENCY:
-			ret = set_accel_odr_osr(dev, val, NULL);
+			ret = set_accel_odr_osr(val, NULL, dev);
```
Comment (mbolivar-nordic) [inline drivers/sensor/bmi270/bmi270.c:581]: I don't think this is a good change.

Having the device pointer first is a common C idiom for emulating object oriented programming. Placing it first is a signal that `dev` is the receiver.

Similarly for the other analogous changes.

File: drivers/sensor/bmi270/Kconfig [hunk 4]
```diff
@@ -5,8 +5,6 @@
 
 config BMI270
 	bool "BMI270 Inertial measurement unit"
-	default y
-	depends on DT_HAS_BOSCH_BMI270_ENABLED
-	select I2C
+	depends on I2C
```
Comment (mbolivar-nordic) [inline drivers/sensor/bmi270/Kconfig:8]: The point of the PR is adding SPI support for the sensor, but you want to add `depends on I2C`?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Overall looks good to me. Some requests.

File: drivers/sensor/bmi270/bmi270.c (hunk 3 above)
Comment (dchat-nordic) [reply drivers/sensor/bmi270/bmi270.c:581]: Ok, I will flip 'dev' to beginning in next push

**Comment by dchat-nordic** [comment]
@MaureenHelm @mbolivar-nordic @galak Please review. I added requested changes

File: drivers/sensor/bmi270/Kconfig [hunk 5]
```diff
@@ -5,8 +5,7 @@
 
 config BMI270
 	bool "BMI270 Inertial measurement unit"
-	default y
-	depends on DT_HAS_BOSCH_BMI270_ENABLED
```
Comment (MaureenHelm) [inline drivers/sensor/bmi270/Kconfig:9]: Don't remove these lines

File: drivers/sensor/bmi270/bmi270.c (hunk 3 above)
Comment (MaureenHelm) [reply drivers/sensor/bmi270/bmi270.c]: This still needs to be done

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please don't change the `struct device *dev` argument position in one commit and then change it back in the next commit. Fixup the original commit instead.

**Comment by dchat-nordic** [comment]
@MaureenHelm I fixed the required `dev` position.

File: drivers/sensor/bmi270/bmi270.c [hunk 6]
```diff
@@ -18,31 +16,54 @@
 
 LOG_MODULE_REGISTER(bmi270, CONFIG_SENSOR_LOG_LEVEL);
 
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "BMI270 driver enabled without any devices"
+#endif
```
Comment (gmarull) [inline drivers/sensor/bmi270/bmi270.c:21]: mostly redundant with DT_HAS* helpers

File: drivers/sensor/bmi270/bmi270.h [hunk 7]
```diff
@@ -206,8 +213,46 @@ struct bmi270_data {
...
+#endif
+
+#if BMI270_BUS_I2C
+extern const struct bmi270_bus_io bmi270_bus_io_i2c;
+#endif
+
+struct bmi270_dev_config {
+	union bmi270_bus bus;
+	const struct bmi270_bus_io *bus_io;
 };
```
Comment (gmarull) [inline drivers/sensor/bmi270/bmi270.h:256]: why is all this in a header file? is it shared with any other module? please move it to the driver otherwise.

File: drivers/sensor/bmi270/bmi270_i2c.c [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Bus-specific functionality for BMI270s accessed via I2C.
+ */
+
+#include "bmi270.h"
+
+#if BMI270_BUS_I2C
```
Comment (gmarull) [inline drivers/sensor/bmi270/bmi270_i2c.c:13]: why not skip compilation of this file if not needed?

File: drivers/sensor/bmi270/bmi270_spi.c [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+ */
+
+/*
+ * Bus-specific functionality for BMI270s accessed via SPI.
+ */
+
+#include <zephyr/logging/log.h>
+#include "bmi270.h"
+
+#if BMI270_BUS_SPI
```
Comment (gmarull) [inline drivers/sensor/bmi270/bmi270_spi.c:14]: ditto. Why isn't that part of the same driver file?

File: drivers/sensor/bmi270/bmi270_spi.c [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+	const struct spi_buf tx_buf[2] = {
+		{.buf = &addr, .len = sizeof(addr)},
+		{.buf = (uint8_t *)data, .len = len}
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf)
+	};
+
+	addr = start & 0x7F;
```
Comment (gmarull) [inline drivers/sensor/bmi270/bmi270_spi.c:77]: magic constant here

File: drivers/sensor/bmi270/bmi270_spi.c [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+	ret = spi_write_dt(&bus->spi, &tx);
+	if (ret) {
+		LOG_DBG("spi_write_dt failed %i", ret);
```
Comment (gmarull) [inline drivers/sensor/bmi270/bmi270_spi.c:81]: `LOG_ERR`?

File: drivers/sensor/bmi270/bmi270_spi.c [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf)
+	};
+
+	addr = start & 0x7F;
+
+	ret = spi_write_dt(&bus->spi, &tx);
+	if (ret) {
```
Comment (gmarull) [inline drivers/sensor/bmi270/bmi270_spi.c:80]: please use expressions evaluating to boolean, e.g. `ret < 0`

**Review by MaureenHelm — DISMISSED** [review]
Need to address twister failures, but otherwise LGTM

File: drivers/sensor/bmi270/bmi270_i2c.c (hunk 8 above)
Comment (mbolivar-nordic) [reply drivers/sensor/bmi270/bmi270_i2c.c:13]: > why not skip compilation of this file if not needed?

That's not possible to do with the current cmake/modules/extensions.cmake, and this is following the pattern for BME280. Unless you have some suggestion for how to do this with existing cmake support for DT?

File: dts/bindings/sensor/bosch,bmi270-spi.yaml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2022, Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    The BMI270 is an inertial measurment unit. See more info at:
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/bosch,bmi270-spi.yaml:5]: You can define this once in bosch,bmi270.yaml and not copy/paste it into the spi and i2c files. Please fix the "measurement" typo.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Looks mainly OK for when the merge window opens. Since we have some time until that happens, I'd like to ask for a few more things. Then I am +1.

**Review by mbolivar-nordic — DISMISSED** [review]
Revoking my -1 since Maureen approved.

File: drivers/sensor/bmi270/bmi270_i2c.c (hunk 8 above)
Comment (dchat-nordic) [reply drivers/sensor/bmi270/bmi270_i2c.c]: After a talk with @kapi-no he suggested a solution. Please check Kconfig file and CMakeLists.txt to see it.

**Review by MaureenHelm — APPROVED** [review]

**Review by gmarull — APPROVED** [review]
